### PR TITLE
Feature/tcvp 2181

### PR DESCRIPTION
**Pull request by lfdgov**
# Description

This PR includes the following proposed change(s):

- subtasks of TCVP-2181 related to relaxing OCR rules

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

OCRd both valid and invalid tickets and searched for handwritten tickets

## Does the change impact or break the Docker build?

- [ ] Yes
- [x] No

If Yes: Has Docker been updated accordingly?

- [ ] Yes
- [ ] No

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Sharkww — APPROVED** [review]

File: src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs [hunk 1]
```diff
@@ -73,12 +76,106 @@ public static void Sanitize(OcrViolationTicket violationTicket)
...
-                violationTicket.Fields[OcrViolationTicket.Count3ActRegs].Value = "MVA";
-                violationTicket.Fields[OcrViolationTicket.Count3Section].Value = count3Section.Replace("MVA", "");
+                string actReg = count3Section.Substring(1, 3);
+                violationTicket.Fields[OcrViolationTicket.Count3ActRegs].Value = actReg;
+                violationTicket.Fields[OcrViolationTicket.Count1Section].Value = count3Section.Replace((actReg is not null ? actReg : ""), "")?.Trim();
             }
         }
+
+        // If any act regs starts with MU replace with MV since the U is likely a mis-scan of the V
+        if (violationTicket.Fields.ContainsKey(OcrViolationTicket.Count1ActRegs))
```
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs:88]: A lot of duplicate code. It is also long and verbose. Hard to follow. This code could be simplifeid like:

```cs
        void ReplaceMUWithMV(string fieldName)
        {
            if (violationTicket.Fields.TryGetValue(fieldName, out var field))
            {
                field.Value = field.Value?.Replace("MU", "MV");
            }
        }

        ReplaceMUWithMV(OcrViolationTicket.Count1ActRegs);
        ReplaceMUWithMV(OcrViolationTicket.Count2ActRegs);
        ReplaceMUWithMV(OcrViolationTicket.Count3ActRegs);
    ```

    Note, it is perfectly valid to have a function declared inside a method.

File: src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs [hunk 2]
```diff
@@ -73,12 +76,106 @@ public static void Sanitize(OcrViolationTicket violationTicket)
...
+                violationTicket.Fields[OcrViolationTicket.ViolationTicketNumber].Value = ticketNumber.Replace("A0", "AO");
+            }
+            if (ticketNumber is not null && ticketNumber.StartsWith("A1"))
+            {
+                violationTicket.Fields[OcrViolationTicket.ViolationTicketNumber].Value = ticketNumber.Replace("A1", "AI");
+            }
+        }
+
+        // Pre-process description and act/regs fields for counts if description number contains text of MVA or MVR strip it out and replace blank act/regs field text
+        if (violationTicket.Fields.ContainsKey(OcrViolationTicket.Count1ActRegs) && violationTicket.Fields.ContainsKey(OcrViolationTicket.Count1Description))
```
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs:111]: The same can be done with these blocks. They all seem the same except for the two field names, 

```
        // Pre-process description and act/regs fields for counts if description number contains text of MVA or MVR strip it out and replace blank act/regs field text
        void ProcessActRegs(string regs, string description)
        {
            if (violationTicket.Fields.ContainsKey(regs) && violationTicket.Fields.ContainsKey(description))
            {
                string? desc = violationTicket.Fields[description].Value;
                bool actRegPop = violationTicket.Fields[regs].IsPopulated();
                if (desc is not null && (desc.Contains("MVA") || desc.Contains("MVR")) && actRegPop == false)
                {
                    string actReg = "MVA";
                    if (desc.Contains("MVR")) actReg = "MVR";
                    violationTicket.Fields[description].Value = violationTicket.Fields[description].Value?.Replace(actReg, "");
                    violationTicket.Fields[regs].Value = actReg;
                }
            }
        }

        ProcessActRegs(OcrViolationTicket.Count1ActRegs, OcrViolationTicket.Count1Description);
        ProcessActRegs(OcrViolationTicket.Count2ActRegs, OcrViolationTicket.Count2Description);
        ProcessActRegs(OcrViolationTicket.Count3ActRegs, OcrViolationTicket.Count3Description);
```

File: src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs (hunk 1 above)
Comment (lfdgov) [reply src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs]: thanks for your feedback -- updated to be a bit more readable hopefully

**Review by KDNeufeld — APPROVED** [review]
Great work.

File: src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs [hunk 3]
```diff
@@ -42,43 +43,104 @@ public async Task ValidateViolationTicketAsync(OcrViolationTicket violationTicke
...
+            SplitSectionActRegs(sectionKey, actRegsKey);
+            ReplaceMUWithMV(actRegsKey);
+            GetActRegsFromDescription(descKey, actRegsKey);
+        }
+
+        SanitizeCount(OcrViolationTicket.Count1Section, OcrViolationTicket.Count1ActRegs, OcrViolationTicket.Count1Description);
+        SanitizeCount(OcrViolationTicket.Count2Section, OcrViolationTicket.Count2ActRegs, OcrViolationTicket.Count2Description);
+        SanitizeCount(OcrViolationTicket.Count3Section, OcrViolationTicket.Count3ActRegs, OcrViolationTicket.Count3Description);
+
+        // Pre-process ticket number if scan reads an A Oh as A Zero replace the Zero with an O in the second position
```
[the remark this answers is not in the document]
Comment (lfdgov) [reply src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs:99]: I added a couple more -- but realistically its kind of random -- only where the digit looks very much like a capital letter, e.g. 0, O or 1, I or 5,S or 8,B

File: src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs [hunk 4]
```diff
@@ -42,43 +43,104 @@ public async Task ValidateViolationTicketAsync(OcrViolationTicket violationTicke
     public static void Sanitize(OcrViolationTicket violationTicket)
     {
         // It can happen that if adjacent text fields has content too close to the common dividing line, the OCR tool can misread both fields thinking one is blank and the other starts with the blank field's text.
+        // If the Acts/Regs section is blank (should be MVA or MVR) and the adjacent Section text starts with "MVA/R" (shouldn't start with MVA/R), then move the MVA/R text to the correct field.
+        void SplitSectionActRegs(string sectionKey, string actRegsKey)
+        {
+            if (violationTicket.Fields.ContainsKey(sectionKey) && violationTicket.Fields.ContainsKey(actRegsKey))
```
[the remark this answers is not in the document]
Comment (lfdgov) [reply src/backend/TrafficCourts/Citizen.Service/Validators/FormRecognizerValidator.cs:50]: I have added a TODO in the code here

**Review by pbolduc — COMMENTED** [review]
Various code clean up could still be done, but it is ok. Was the unit test fixed?
